### PR TITLE
Refactor&Feature: templatize `ESolver_KS_LCAO` , unify `ops/opsd`, `psi/psid`; real-valued Davidson

### LINKED ISSUE #2951: Unify numerical type template in PW calculation
### Background

The usage of template parameters in ABACUS is relatively confusing. Here are a few examples:

- `complex<FPTYPE>` is used uniformly in PW code, that is, `FPYTPE` cannot take `complex<double>`;

- There is `complex<T>` in the DiagoBPCG written by @denghuilu, so `T` here cannot be `complex<double>` too;

- However: `T` in `HamiltonLCAO<T>` can be `complex<double>`; 

- And In the new `ESolver_LR` module, `FPTYPE` can also be `complex<double>`... 

### Describe the solution you'd like

In the PW calculation, we employ a unified numerical type template represented by 'T' instead of 'FPTYPE' to denote the calculation's numerical type. This 'T' template is designed to accommodate various data types, including 'std::complex<double>' and others.

### Additional Context

_No response_

### Task list for Issue attackers

- [x] Review and understand the proposed feature and its importance.
- [x] Research on the existing solutions and relevant research articles/resources.
- [x] Discuss with the team to evaluate the feasibility of implementing the feature.
- [x] Create a design document outlining the proposed solution and implementation details.
- [x] Get feedback from the team on the design document.
- [x] Develop the feature following the agreed design.
- [ ] Write unit tests and integration tests for the feature.
- [ ] Update the documentation to include the new feature.
- [ ] Perform code review and address any issues.
- [ ] Merge the feature into the main branch.
- [ ] Monitor for any issues or bugs reported by users after the feature is released.
- [ ] Address any issues or bugs reported by users and continuously improve the feature.

### PR DESCRIPTION
**Pull request by maki49**
### Refactor
fix #2951 (achieving the final goal)
Thanks for the previous refactor by @denghuilu  and @dyzheng. Now it's much more convenient for me to implement a new ESolver. 
### Feature (based on the refactor)
support real(double)-DiagoDavid for #2460 , by refactoring some **math kernels** (including CUDA and HIP) in module_hsolver, and templatizing some tool-funcs

### CODE REVIEW

**Review by dyzheng — APPROVED** [review]

File: source/module_hsolver/diago_david.cpp [hunk 1]
```diff
@@ -781,7 +784,7 @@ void DiagoDavid<T, Device>::refresh(const int& dim,
     basis.zero_out();
     for (int m = 0; m < nband; m++)
     {
-        syncmem_complex_op()(this->ctx, this->ctx, &basis(m, 0), &psi(m, 0), this->dim);
+        syncmem_complex_op()(this->ctx, this->ctx, &basis(m, 0), psi.get_k_first() ? &psi(m, 0) : &psi(m, 0, 0), this->dim);
```
Comment (dyzheng) [inline source/module_hsolver/diago_david.cpp:795]: Can you refactor Psi::fix_k() to skip k index without k_first?

Comment (maki49) [reply source/module_hsolver/diago_david.cpp:795]: No need to skip. k_first=0 means that nks is more like a dimension of **basis**  to be contracted in "HPsi". In LR-TDDFT the formula writes: 
$$\sum_{jb\mathbf{k}'}A^I_{ia\mathbf{k}, jb\mathbf{k}'}X^I_{jb\mathbf{k}'}$$
In the code: 
-  H means A
- Psi means X
- "band" means the superscript I : the number of excited states to be solved
- k: k-points, the same meaning as the ground state
- "basis": number of occupied ks-orbitals (subscripts i,j) * number of unoccupied ks-orbitals (subscripts a,b), corresponding to "bands" of the ground state

Comment (dyzheng) [reply source/module_hsolver/diago_david.cpp:795]: > No need to skip. k_first=0 means that nks is a dimension of **basis** to be contracted in "HPsi". In LR-TDDFT the formula writes: ∑jbk′Aaik,jbk′IXjbk′I In the code:
> 
> * H means A
> * Psi means X
> * "band" means the superscript I (the number of excited states to be solved)
> * k: k-points, same meaning as the ground state
> * "basis": number of occupied ks-orbitals (subscripts i,j) * number of unoccupied ks-orbitals (subscripts a,b), corresponding to "bands" of the ground state

Can you add these comments into this code file? I think other developers will also confused with k_first related code.

Comment (dyzheng) [reply source/module_hsolver/diago_david.cpp:795]: In principle, DiagoDavid only do diagonalization for target Psi which contains matrix with size of nbands * nbasis.
For LR-TDDFT, the better practice would be use a new Psi with 1 * nbands * (nks*nbasis) as a wrapper of current Psi with nbands * nks * nbands. 
The DiagoDavid class should not treat different DFT algorithm, it is only a basic math tool.

Comment (maki49) [reply source/module_hsolver/diago_david.cpp:795]: I've added them at the place setting "dim" in DiagoDavid.

Comment (maki49) [reply source/module_hsolver/diago_david.cpp:795]: > In principle, DiagoDavid only do diagonalization for target Psi which contains matrix with size of nbands * nbasis. For LR-TDDFT, the better practice would be use a new Psi with 1 * nbands * (nks*nbasis) as a wrapper of current Psi with nbands * nks * nbands. The DiagoDavid class should not treat different DFT algorithm, it is only a basic math tool.

I think `DiagoDavid` is not a basic math tool unless it does not depend on the class `Hamilt` and `Psi`. Now it depends on `Psi`, then using any logic implemented in `Psi` is reasonable. 
If in the future we need to refactor `Diago` class with no `Hamilt` and `Psi` in the interface (such as HPC-friendly pointer-only interface),  the k-first logic would be naturally moved out of `DiagoDavid`.

**Review by dyzheng — APPROVED** [review]
please solve the conflicts.

**Comment by dyzheng** [comment]
Please solve conflicts again due to #2998

**Review by denghuilu — APPROVED** [review]
LGTM

**Pull request merged**